Instructor: Question view: visibility icon does not appear for new comments #8899

nidhi98gupta opened the pull request:
Fixes #8899 

Outline of solution
- Removed `hasVisibilityIcon`
- Eye icon will now appear for all the comments
-  if comment is not visible to anyone we show  `This comment is shown to nobody`

tshradheya commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackResponseCommentAjaxPageData.java:
```diff
@@ -40,7 +40,12 @@ public FeedbackResponseCommentRow getComment() {
                                                getResponseVisibilities(), instructorEmailNameTable, sessionTimeZone);
         frc.enableEditDelete();
-
+        String whoCanSeeComment;
```
Declare this in line 46?

Keep the scope as small as possible. Since we only need it in the `if` part, declare and initialize there itself

nidhi98gupta commented:
@tshradheya Done. Ready for review.

xpdavid commented on src/main/java/teammates/ui/pagedata/InstructorFeedbackResponseCommentAjaxPageData.java:
```diff
@@ -41,6 +41,11 @@ public FeedbackResponseCommentRow getComment() {
                                                getResponseVisibilities(), instructorEmailNameTable, sessionTimeZone);
         frc.enableEditDelete();
 
+        boolean isVisibilityIconShown = !frc.getShowCommentToString().isEmpty();
```
Add a comment
```
// only show icon if the comment is visible to at least someone
```

On a separate note, shall we just remove the constrain for all `results.feedbackSession.isPublished()`?

nidhi98gupta replied:
Yes, Should I do that in this PR or open a new one for it?

nidhi98gupta commented:
I think eye icon can be shown at all times. Even if comment is not visible to anyone we show something like `This comment is shown to nobody`?

tshradheya commented:
> I think eye icon can be shown at all times. Even if comment is not visible to anyone we show something like This comment is shown to nobody?

Sounds good. If this change is not very tough to implement lets do it in this PR itself and change the scope of PR

nidhi98gupta commented:
@xpdavid @tshradheya I am unable to make god mode changes to `InstructorSearchPageUiTest.java`, this test fails for me on each one of my brach including `master` with stack trace being
```
java.time.format.DateTimeParseException: Text '2026-03-01 11:59 PM +0000' could not be parsed at index 10
	at java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:1949)
	at java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1851)
	at java.time.Instant.parse(Instant.java:395)
	at teammates.common.util.JsonUtils$TeammatesInstantAdapter.deserialize(JsonUtils.java:88)
	at teammates.common.util.JsonUtils$TeammatesInstantAdapter.deserialize(JsonUtils.java:79)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.Gson.fromJson(Gson.java:927)
	at com.google.gson.Gson.fromJson(Gson.java:892)
	at com.google.gson.Gson.fromJson(Gson.java:841)
	at teammates.common.util.JsonUtils.fromJson(JsonUtils.java:66)
	at teammates.storage.search.FeedbackResponseCommentSearchDocument.fromResults(FeedbackResponseCommentSearchDocument.java:259)
	at teammates.storage.api.FeedbackResponseCommentsDb.search(FeedbackResponseCommentsDb.java:314)
	at teammates.logic.core.FeedbackResponseCommentsLogic.searchFeedbackResponseComments(FeedbackResponseCommentsLogic.java:160)
	at teammates.logic.api.Logic.searchFeedbackResponseComments(Logic.java:1983)
	at teammates.ui.controller.InstructorSearchPageAction.execute(InstructorSearchPageAction.java:51)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:476)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:75)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:50)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
```
It's strange that it only fails on my system. I tried investigating and I think it is because `2026-03-01T23:59:00Z` gets converted to `2026-03-01 11:59 PM +0000` when saving to Search Document and while converting it back to `Instant`, it fails. What can I do to resolve this?

xpdavid commented:
We migrate the time format some time in this year I think. You could just delete the local datastore (`appengine-generated/local_db.bin`) if you don't mind (Or to run `DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat` this script in `client` package).

nidhi98gupta commented:
@xpdavid I did both of the things, but it still fails.

xpdavid commented:
You need change the `isPreview()` to false in DataMigrationForFeedbackResponseCommentSearchDocumentDateFormat. In addition, there are two `appengine-generated` folders, I hope you delete both.

nidhi98gupta commented:
@xpdavid I changed `isPreview` to false.
```
Preview: false
Total number of feedbackresponsecomments: 38
Number of affected feedbackresponsecomments: 0
Number of updated feedbackresponsecomments: 0
--- Remote operation completed --
```
I think there weren't two `appengine-generated` folders.  I deleted `teammates/WEB-INF` folder which contained `appengine-generated` folder. Test still failed.

xpdavid commented:
If you delete the folder, why there is still instance in the db?

> Total number of feedbackresponsecomments: 38

nidhi98gupta commented on src/main/java/teammates/ui/template/FeedbackResponseRow.java:
```diff
@@ -98,13 +88,17 @@ public String getResponseText() {
      */
     public String getTypeOfPeopleCanViewComment(FeedbackResponseCommentAttributes comment,
                                                 FeedbackQuestionAttributes relatedQuestion) {
```
This function is duplicated in `PageData` but it can be used it here because of macker rule.

nidhi98gupta replied:
Or put it in constructor? For that we can pass `FeedbackQuestionAttributes question` through constructor and then use `getTypeOfPeopleCanViewComment`?

xpdavid replied:
Also can if it is compulsory to have this setting (I think yes as we now remove the checking).

nidhi98gupta commented:
@xpdavid  @tshradheya Ready for review.

nidhi98gupta commented:
@xpdavid Ready for review

xpdavid commented on src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java:
```diff
@@ -240,9 +236,8 @@ public void enableEditDelete() {
-    public void setVisibilityIcon(boolean hasVisibilityIcon, String whoCanSeeComment) {
-        this.hasVisibilityIcon = hasVisibilityIcon;
-        this.whoCanSeeComment = whoCanSeeComment;
+    public void setVisibilityIconString(String visibilityIconString) {
```
Do we need the setter?

nidhi98gupta replied:
Not needed. Removing it.

nidhi98gupta commented:
@xpdavid @tshradheya Ready to merge :D

damithc commented on src/main/java/teammates/ui/template/FeedbackResponseCommentRow.java:
```diff
@@ -267,4 +258,57 @@ private String getEditedAtText(String lastEditorEmail, Instant createdAt, Instan
+     * Returns the type of people that can view the response comment.
+     */
+    private String getTypeOfPeopleCanViewComment(FeedbackResponseCommentAttributes comment,
+                                                FeedbackQuestionAttributes relatedQuestion) {
```
Shouldn't this be 8-space indentation?

damithc commented on src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java:
```diff
@@ -193,7 +193,7 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
             String answer = response.getResponseDetails().getAnswerHtmlStudentView(questionDetails);
             List<FeedbackResponseCommentRow> comments = createStudentFeedbackResultsResponseComments(
-                                                                                          response.getId());
+                                                                                          response.getId(), question);
```
Could be like this?
```java
List<FeedbackResponseCommentRow> comments = 
        createStudentFeedbackResultsResponseComments(response.getId(), question);
```

nidhi98gupta commented:
@tshradheya can merge this?
